examples/ct/ct_handlers.go

```diff
@@ -494,3 +420,64 @@ func marshalLogIDAndSignatureForResponse(sct ct.SignedCertificateTimestamp, km c
 
 	return logID, signature, nil
 }
+
+// buildLeafProtoForAddChain is also used by add-pre-chain and does the hashing to build a
+// LeafProto that will be sent to the backend
+func buildLeafProtoForAddChain(merkleLeaf ct.MerkleTreeLeaf, certChain []*x509.Certificate) (trillian.LeafProto, error) {
+	var leafBuffer bytes.Buffer
+	if err := WriteMerkleTreeLeaf(&leafBuffer, merkleLeaf); err != nil {
+		glog.Warningf("Failed to serialize merkle leaf: %v", err)
+		return trillian.LeafProto{}, err
+	}
+
+	var logEntryBuffer bytes.Buffer
+	logEntry := NewCTLogEntry(merkleLeaf, certChain)
+	if err := logEntry.Serialize(&logEntryBuffer); err != nil {
+		glog.Warningf("Failed to serialize log entry: %v", err)
+		return trillian.LeafProto{}, err
+	}
+
+	// leafHash is a crosscheck on the data we're sending in the leaf buffer. The backend
+	// does the tree hashing.
+	leafHash := sha256.Sum256(leafBuffer.Bytes())
+
+	return trillian.LeafProto{LeafHash: leafHash[:], LeafData: leafBuffer.Bytes(), ExtraData: logEntryBuffer.Bytes()}, nil
+}
+
+// marshalAndWriteAddChainResponse is used by add-chain and add-pre-chain to create and write
+// the JSON response to the client
+func marshalAndWriteAddChainResponse(sct ct.SignedCertificateTimestamp, km crypto.KeyManager, w http.ResponseWriter) error {
+	logID, signature, err := marshalLogIDAndSignatureForResponse(sct, km)
+
+	if err != nil {
+		glog.Warningf("failed to marshal for response: %v", err)
+		return err
+	}
+
+	resp := addChainResponse{
+		SctVersion: int(sct.SCTVersion),
+		Timestamp:  sct.Timestamp,
+		ID:         base64.StdEncoding.EncodeToString(logID[:]),
+		Extensions: "",
+		Signature:  signature}
+
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	jsonData, err := json.Marshal(&resp)
+
+	if err != nil {
+		glog.Warningf("Failed to marshal add-chain resp: %v", resp)
+		sendHttpError(w, http.StatusInternalServerError, err)
+		return err
+	}
+
+	_, err = w.Write(jsonData)
```

Comment:
The number of bytes written. The docs say writers must return an error if they write less than the data supplied.